### PR TITLE
VS: reduce some coupling between components

### PR DESCRIPTION
**Pull request by majocha**
Especially StatusBar was passed around waaay too much.

Also: grab VS services on the main thread just in case.

### CODE REVIEW

**Review by psfinaki — COMMENTED** [review]
@majocha thanks for your efforts to reduce the chaos in the Editor :)

I am still getting my head around this, but for now let me ask you this (maybe I misunderstand something). 

Looking at your changes, I see we are getting rid of some parameters passed to some services, replacing them with calls to `GlobalProvider`. Are you sure it's a move to the right direction? Propagating all the same parameters through the code flow is not good but this feels a bit like Service Locator which is arguably not a nice thing either.

I know this is happening in some other places in code but I wonder if this is the consistency we want to achieve.

---
P.S. A small ask for the future. This is definitely a PR of an acceptable size - still it would benefit from a guiding description to the contributors - like where to start and so on. Although I hope AI will soon be able to generate this for us :)

**Comment by majocha** [comment]
@psfinaki this is mostly to unblock me on the quickinfo tests in the other PR.
I tried to follow what [VisualStudio Toolkit](https://github.com/VsixCommunity/Community.VisualStudio.Toolkit) does.
Namely, calling GetService in VS requires main thread, and as I remember this was a source of pain and bugs historically. Centralizing those calls helps with keeping this in check.

As I look at this again, I think keeping extension methods on IServiceProvider would be good enough, as GlobalProvider is always available from VS shell. I'll modify this.

**Comment by psfinaki** [comment]
Alright. Yep, thanks for the explanation. Eventually, it should be exactly this - moving towards fewer bugs and better testability.

**Comment by majocha** [comment]
I left Shell's GlobalProvider where it made the most sense and modified extension methods on IAsyncServiceProvider and IServiceProvider to resemble what [Roslyn ](https://github.com/dotnet/roslyn/blob/5f135a0ff24ed391207152a3f217c9f68a9449d1/src/EditorFeatures/Core.Wpf/Utilities/RoslynServiceExtensions.cs)does.

File: vsintegration/src/FSharp.Editor/Navigation/GoToDefinition.fs [hunk 1]
```diff
@@ -109,7 +109,10 @@ module private ExternalSymbol =
         | _ -> []
 
 // TODO: Uncomment code when VS has a fix for updating the status bar.
```
Comment (psfinaki) [inline vsintegration/src/FSharp.Editor/Navigation/GoToDefinition.fs:111]: BTW do you think this is still relevant? :)

File: vsintegration/src/FSharp.Editor/Navigation/GoToDefinition.fs [hunk 2]
```diff
@@ -818,6 +811,7 @@ type internal FSharpNavigation
 
     member _.TryGoToDefinition(position, cancellationToken) =
         let gtd = GoToDefinition(metadataAsSource)
+        let statusBar = StatusBar()
```
Comment (psfinaki) [inline vsintegration/src/FSharp.Editor/Navigation/GoToDefinition.fs:814]: So, here (and elsewhere) - are you sure creating a status bar from scratch is equivalent to passing around the thing that flows in?

File: vsintegration/src/FSharp.Editor/Common/Vs.fs [hunk 3]
```diff
@@ -91,9 +92,36 @@ module internal VsRunningDocumentTable =
 
 [<AutoOpen>]
 module internal ServiceProviderExtensions =
-    type internal System.IServiceProvider with
```
Comment (psfinaki) [inline vsintegration/src/FSharp.Editor/Common/Vs.fs:94]: So once again, what's the need for this change? I'm not against it, just a bit paranoid and trying to better understand all this magic

File: vsintegration/src/FSharp.Editor/Navigation/GoToDefinition.fs (hunk 1 above)
Comment (majocha) [reply vsintegration/src/FSharp.Editor/Navigation/GoToDefinition.fs:111]: 😀 I bet VS statusbar works fine. Just not in our code. We should really stand of the shoulders of experts and either copy what [VSIX Community Toolkit](https://github.com/VsixCommunity/Community.VisualStudio.Toolkit) does or even just install their nuget.

File: vsintegration/src/FSharp.Editor/Navigation/GoToDefinition.fs (hunk 2 above)
Comment (majocha) [reply vsintegration/src/FSharp.Editor/Navigation/GoToDefinition.fs:814]: Looking again at this, IMV we should just yeet all the statusbar code. It does nothing anyway.

Comment (majocha) [reply vsintegration/src/FSharp.Editor/Navigation/GoToDefinition.fs:814]: Looks like it was just a clunky wrapper around SVsStatusbar service, when it worked, pretty much stateless on its own, so creating it ad hoc is ok. Removing this completely looks like serious work, this file is 1000+ freaking lines🙃

**Comment by majocha** [comment]
@psfinaki, I'll draft this, since I managed to fix my quickinfo pr with minimal changes and this is not urgently needed.

**Comment by psfinaki** [comment]
@majocha actually it feels like we are getting to the point :) If you've noticed that StatusBar is useless, feel free to create a ticket about that and we can jump on it - either from this PR or from scratch. But this will be a major relief - the more crap we remove from the Editor code the better it is for everyone!

Meanwhile thanks for all the investigation, I'm also learning a lot here.

**Comment by majocha** [comment]
@psfinaki, makes sense, I'll make an issue.

**Pull request closed without merging**